csrc/dynamic_transform.cpp

```diff
@@ -376,13 +380,76 @@ void DynamicTransformConcretizer::concretize() {
   // Set output IterTypes for dynamic resize ops
   concretizeResize();
 
-  // Finally, propagate concretized domains
-  auto all_stmts = StmtSort::getStmts(info_->fusion(), true);
-  for (auto stmt : all_stmts) {
-    if (stmt->isA<Val>()) {
-      mutate(stmt);
+  // This fixes the set of statements we will process over the course of
+  // multiple passes.
+  // Since we will traverse these Statements after some have been removed, we
+  // will not be able to safely check the types of each Statement in later
+  // loops. To avoid segfaults, we first split all_statements into subsets for
+  // each traversal.
+  std::vector<Val*> non_tds_tvs;
+  std::vector<Expr*> all_exprs;
+  std::vector<Val*> tvs_and_tds;
+  for (auto stmt : StmtSort::getStmts(info_->fusion(), true)) {
+    if (stmt->isExpr()) {
+      all_exprs.push_back(stmt->asExpr());
+    } else {
+      auto val = stmt->asVal();
+      if (val->isA<TensorView>() || val->isA<TensorDomain>()) {
+        tvs_and_tds.push_back(val);
+      } else {
+        non_tds_tvs.push_back(val);
+      }
```

Comment:
What was previously a single loop over `all_stmts` is now three separate loops over these subsets.